src/libcrun/linux.c

```diff
@@ -1947,6 +1947,82 @@ do_mounts (libcrun_container_t *container, int rootfsfd, const char *rootfs, con
   return 0;
 }
 
+/*
+ * libcrun_container_do_bind_mount
+ *
+ *  Allows external plugins and handlers to perform bind `mounts` on container.
+ *  returns: 0 if successful anything else states `error` and configures `err` with relevent error.
+ */
+int
+libcrun_container_do_bind_mount (libcrun_container_t *container, char *mount_source, char *mount_destination, char **mount_options, size_t mount_options_len, libcrun_error_t *err)
+{
+  int ret, rootfsfd;
+  size_t rootfs_len = get_private_data (container)->rootfs_len;
+
+  const char *target = consume_slashes (mount_destination);
+  cleanup_free char *data = NULL;
+  unsigned long flags = 0;
+  unsigned long extra_flags = 0;
+  cleanup_close int targetfd = -1;
+  int is_dir = 1;
+  uint64_t rec_clear = 0;
+  uint64_t rec_set = 0;
+  cleanup_close int rootfsfd_cleanup = -1;
+  const char *rootfs = get_private_data (container)->rootfs;
+
+  if (rootfs == NULL)
+    rootfsfd = AT_FDCWD;
+  else
+    {
+      rootfsfd = rootfsfd_cleanup = open (rootfs, O_PATH | O_CLOEXEC);
+      if (UNLIKELY (rootfsfd < 0))
+        return crun_make_error (err, errno, "open `%s`", rootfs);
+
+      // since we have opened rootfsfd again update container private data
+      // otherwise we get `crun: invalid rootfs state` error later while
+      // performing mount.
+      get_private_data (container)->rootfsfd = rootfsfd;
```

Comment:
this is risky.

We do not reset its value and also we override the previous one if it was a valid fd

Reply:
Could I check if its valid fd then not replace it but only replace if its invalid `fd` ? Cause I'm re-using `do_mount(` and it checks it again here https://github.com/containers/crun/blob/main/src/libcrun/linux.c#L831 at some point.

Reply:
we would leave an invalid fd.  I think we should not try to override its value.

Maybe in addition to `HANDLER_CONFIGURE_BEFORE_MOUNTS` and `HANDLER_CONFIGURE_AFTER_MOUNTS` we need also `HANDLER_CONFIGURE_MOUNTS`
and we return an error when rootfsfd is not a valid fd.

Reply:
Yes I agree checking this new state `HANDLER_CONFIGURE_MOUNTS` and will open a seperate PR for this new state.

Reply:
Added a new commit.